drivers/i2s/i2s_cavs.c

```diff
@@ -0,0 +1,791 @@
+		break;
+
+	default:
+		SYS_LOG_ERR("Unsupported Clock format");
+		return -EINVAL;
+	}
+
+	ssc0 |= SSCR0_MOD | SSCR0_ACS;
+
+	mclk = SOC_INTEL_S1000_MCK_XTAL_FREQ_HZ;
+	bit_clk_freq = i2s_cfg->frame_clk_freq * word_size_bits * num_words;
+
+	/* BCLK is generated from MCLK - must be divisible */
+	if (mclk % bit_clk_freq) {
+		SYS_LOG_ERR("MCLK is not a factor of BCLK");
+		return -EINVAL;
+	}
+
+	/* divisor must be within SCR range */
+	mdiv = (mclk / bit_clk_freq) - 1;
+	if (mdiv > (SSCR0_SCR_MASK >> 8)) {
+		SYS_LOG_ERR("Divisor is not within SCR range");
+		return -EINVAL;
+	}
+
+	/* set the SCR divisor */
+	ssc0 |= SSCR0_SCR(mdiv);
+
+	/* word_size_bits must be <= 38 for SSP */
+	if (word_size_bits > 38) {
+		SYS_LOG_ERR("word size bits is more than 38");
+		return -EINVAL;
+	}
+
+	/* format */
+	switch (i2s_cfg->format & I2S_FMT_DATA_FORMAT_MASK) {
+
+	case I2S_FMT_DATA_FORMAT_I2S:
+		ssc0 |= SSCR0_FRDC(i2s_cfg->channels);
+
+		/* set asserted frame length */
+		frame_len = word_size_bits;
+
+		/* handle frame polarity, I2S default is falling/active low */
+		sspsp |= SSPSP_SFRMP(!inverted_frame);
+		break;
+
+	case I2S_FMT_DATA_FORMAT_LEFT_JUSTIFIED:
+		ssc0 |= SSCR0_FRDC(i2s_cfg->channels);
+
+		/* LJDFD enable */
+		ssc2 &= ~SSCR2_LJDFD;
+
+		/* set asserted frame length */
+		frame_len = word_size_bits;
+
+		/* LEFT_J default is rising/active high, opposite of I2S */
+		sspsp |= SSPSP_SFRMP(inverted_frame);
+		break;
+
+	case I2S_FMT_DATA_FORMAT_PCM_SHORT:
+	case I2S_FMT_DATA_FORMAT_PCM_LONG:
+	default:
+		SYS_LOG_ERR("Unsupported I2S data format");
+		return -EINVAL;
+	}
+
+	sspsp |= SSPSP_SFRMWDTH(frame_len);
+
+	if (word_size_bits > 16) {
+		ssc0 |= (SSCR0_EDSS | SSCR0_DSIZE(word_size_bits - 16));
+	} else {
+		ssc0 |= SSCR0_DSIZE(word_size_bits);
+	}
+
+	ssp->ssc0 = ssc0;
+	ssp->ssc1 = ssc1;
+	ssp->ssc2 = ssc2;
+	ssp->ssc3 = ssc3;
+	ssp->sspsp2 = sspsp2;
+	ssp->sspsp = sspsp;
+	ssp->ssioc = ssioc;
+	ssp->ssto = ssto;
+	ssp->sstsa = sstsa;
+	ssp->ssrsa = ssrsa;
+
+	/* Set up DMA channel parameters */
+	word_size_bytes = (word_size_bits + 7) / 8;
+	strm->dma_cfg.source_data_size = word_size_bytes;
+	strm->dma_cfg.dest_data_size = word_size_bytes;
+
+	strm->state = I2S_STATE_READY;
+	return 0;
+}
+
+static int tx_stream_start(struct stream *strm, struct ssp *const ssp,
+			  struct device *dev_dma)
+{
+	size_t mem_block_size;
+	int ret;
+
+	ret = queue_get(&strm->mem_block_queue, strm->cfg.options,
+			&strm->mem_block, &mem_block_size);
+	if (ret < 0) {
+		return ret;
+	}
+	k_sem_give(&strm->sem);
+
+	/* Assure cache coherency before DMA read operation */
+	DCACHE_CLEAN(strm->mem_block, mem_block_size);
+
+	ret = start_dma(dev_dma, strm->dma_channel, &strm->dma_cfg,
+			strm->mem_block, (void *)&(ssp->ssd),
+			mem_block_size);
+	if (ret < 0) {
+		SYS_LOG_ERR("Failed to start TX DMA transfer: %d", ret);
+		return ret;
+	}
+
+	/* enable port */
+	ssp->ssc0 |= SSCR0_SSE;
+
+	/* enable DMA */
```

Comment:
So... start_dma() doesn't enable DMA?  Is that racy?   Seems like this needs a comment explaining why this is OK.  Maybe "start_dma()" should be renamed "setup_dma()"?

Reply:
Well.. here the DMA is controlled by the peripheral. So, even if the DMA starts, unless the peripheral enables the DMA, the DMA will not complete. I can put in a comment there.